python/cudf/cudf/utils/utils.py

```diff
@@ -371,3 +372,34 @@ def _cudf_nvtx_annotate(func, domain="cudf_python"):
 _dask_cudf_nvtx_annotate = partial(
     _cudf_nvtx_annotate, domain="dask_cudf_python"
 )
+
+
+class CachedInstanceMeta(type):
+    """
+    Metaclass that caches `maxsize` instances.
+    """
+
+    def __new__(self, names, bases, attrs, maxsize=128):
+        self.__maxsize = maxsize
+        return type.__new__(self, names, bases, attrs)
+
+    def __init__(self, names, bases, attrs, **kwargs):
+        self.__instances = OrderedDict()
```

Comment:
Is there a reason to use an `OrderedDict` rather than just a `dict` (which is insertion ordered as of 3.7)? Is it just so that we can use `popitem` below? `OrderedDict` is slower than `dict` so we should make sure that the cost is worth paying.

Reply:
Yes -- that is still a concern and one I'm still not sure of the best way to handle.

Reply:
> Is it just so that we can use popitem below

There's also a convenient `.move_to_end()` method that is useful in writing LRU caches.

Reply:
>  accidentally keeping a pool alive with cached objects that are otherwise unreferenced.

I actually realized that AFAICT `Scalar` objects don't keep a reference to the memory resource they use. This means  that if RMM is reinitialized, any `Scalar` instances cached prior to the reinitialization may be unsafe to use. Will try to trigger a segfault to confirm.